core/lib/Cardano/Script.hs

```diff
@@ -51,30 +55,28 @@ import qualified Data.List as L
 --
 -- @since 3.0.0
 data Script =
-      RequireSignatureOf KeyHash
-    | RequireAllOf [Script]
-    | RequireAnyOf [Script]
-    | RequireMOf Word8 [Script]
+      RequireSignatureOf !KeyHash
+    | RequireAllOf ![Script]
+    | RequireAnyOf ![Script]
+    | RequireMOf Word8 ![Script]
     deriving stock (Generic, Show, Eq)
-
 instance NFData Script
 
--- | A 'KeyHash' type represents verification key hash
--- that participate in building MultisigScript. The hash is expected to have size of
+-- | A 'ScriptHash' type represents script hash. The hash is expected to have size of
 -- 28-byte.
 --
 -- @since 3.0.0
-newtype KeyHash = KeyHash ByteString
+newtype ScriptHash = ScriptHash ByteString
```

Comment:
Alternatively, we could reuse the `Depth` data-kind and have a `Hash (purpose :: Depth)` data-type. That'd be quite consistent with the approach for keys. So you'd have:

```hs
Shelley 'AddressK XPub -> Hash 'AddressK 
Shelley 'StakingK XPub -> Hash 'StakingK
```

Reply:
I would argue that as ScriptHash is computed based on script which has rulings plus verification key hashes, then maybe it is good as it is. No need to add differentiation. The same ScriptHash can go to payment and stake credential. We captured differentiation already in Credential (purpose::Depth)